### ISSUE TITLE
Problems with non-string arguments when using `BuildIn.run_keyword`

### ISSUE THREAD

**Issue by robinmackaij** [robotframework/robotframework#5031]
When using `BuildIn.run_keyword` the (positional) arguments that follow the target keyword name should be passed to the keyword "as is".

A change in RF7 changed this behavior in a specific situation; if there are (exactly) 2 positional arguments following the keyword name and the first of those arguments is iterable and the second a mapping, the arguments are no longer passed "as is" but they are expanded before being passed to the target keyword.

**Comment by pekkaklarck** [comment]
This is due to #5000. The motivation of that enhancement was making it more convenient for listeners and model modifiers to set keyword arguments, and it didn't occur to me that it would also affect using `BuiltIn.run_keyword` programmatically. Or possibly it did occur, but I thought it's so unlikely that it would cause problems that we don't need to care.

Fixing this issue so that `BuiltIn.run_keyword` works like earlier, without special cases, when used programmatically is easy. The problem is that then libraries using it cannot benefit from the more convenient API provided by #5000. That's unfortunate, but I believe that behavior is safer. We also already have various issues related to making it easier for libraries to call other keywords. I hope we could enhance that situation in general in RF 7.1 or latest in RF 8.0.

**Comment by pekkaklarck** [comment]
This problem can be reproduced also without a library by passing keyword name and arguments as a list variable to `Run Keyword` so that it needs to resolve also arguments:
```
${name and args} =    Evaluate    'Log Many', ['a', 'b', 'c'], {'d': 'e'}
Run Keyword    @{name and args}
```

**Comment by pekkaklarck** [comment]
We discussed about this on Slack some time ago and made a decision to revert the change done in #5000 in RF 7.0.1. Development of that release has been stalled due to my paternity leave, but I try to get it done now and hopefully get also this issue resolved soon.

The original motivation in #5000 was fine and I'll submit a new issue about that. I also have an idea about a less hacky and problematic solution than what was done in #5000.

**Comment by pekkaklarck** [comment]
RF 7.0.1 rc 1 with this fix [has just been released](https://github.com/robotframework/robotframework/blob/master/doc/releasenotes/rf-7.0.1rc1.rst). It would be great if you could test does the fix work properly before the final release that's targeted for Monday June 10.

**Comment by robinmackaij** [comment]
Running the tests of https://github.com/MarketSquare/robotframework-openapitools (more specifically https://github.com/MarketSquare/robotframework-openapitools/blob/main/tests/driver/suites/load_from_url.robot) with the 7.0.1 RC1 yields a new issue: `ValueError: Argument 'status_code' got value '('status_code', '422')' (tuple) that cannot be converted to integer.`

I could not reproduce this with a regular data-driven (test template) test with similar data.

What is interesting is that when I breakpoint within the keyword used as the test template and inspect the test and local scope, I can see that the data is as expected in the `DataDriver_TEST_DATA` dict (test scope) while the key-value pairs have been converted to tuples in the local scope:
![image](https://github.com/robotframework/robotframework/assets/45491813/9441908f-445e-4b31-aa40-a07a12d58a71)

**Comment by pekkaklarck** [comment]
Based on the discussion on Slack, the above problem occurs because DataDriver has already taken the functionality introduced in #5000 into use. That functionality has now been reverted as part of this issue and DataDtriver obviously doesn't work correctly anymore. Being able to use non-string named arguments is important for DataDriver, so we need to enable that functionality somehow. We discussed that on Slack as well and this seems to be the best approach:

- Add _some_ way to use non-string named arguments in RF 7.0.1. That API can be private or even somewhat hacky, the important thing is that DataDriver, and possible other tools needing this functionality, can use it in RF 7.0.1. We already discussed about one possible approach and I'll write a separate comment about it a bit later.
- Add a better public API in RF 7.1. It would be obviously great to add it already in 7.0.1, but the changes are likely so big that there are risks for further regressions. Better to design and test it properly as part of 7.1. I'll submit a separate issue about this.

**Comment by pekkaklarck** [comment]
The new, temporary API for creating named arguments with non-string values has been added. There exists new `robot.running.model.Argument` class accepting `name` and `value`, that can be used in `robot.running.model.Keyword.args`. `Argument` represents a named argument if `name` is a string and a positional argument if `name` is `None`. Tests added as part of 570ff7135988537b9e3a4fd6972f2d0367adea65 show how it can be used with model modifiers, but the same approach works also with listeners and other programmatic usages.

The main limitation of the new API is that it isn't compatible with the JSON model. That's why I believe we should add a better public API and then deprecate this one. The new API will certainly be preserved in RF 7.1, though.

I'll reopen this issue until we get feedback about the new API and until I have submitted an issue about the better API to be added in RF 7.1.
